src/DependencyInjection.Attributed/IncrementalGenerator.cs

```diff
@@ -32,30 +31,72 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
             return visitor.TypeSymbols;
         });
 
-        var selector = (AttributeData attr) =>
+        bool IsService(AttributeData attr) =>
             (attr.AttributeClass?.Name == "ServiceAttribute" || attr.AttributeClass?.Name == "Service") &&
             attr.ConstructorArguments.Length == 1 &&
             attr.ConstructorArguments[0].Kind == TypedConstantKind.Enum &&
             attr.ConstructorArguments[0].Type?.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat) == "global::Microsoft.Extensions.DependencyInjection.ServiceLifetime";
 
+        bool IsExport(AttributeData attr)
+        {
+            var attrName = attr.AttributeClass?.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
+            return attrName == "global::System.Composition.ExportAttribute" ||
+                attrName == "global::System.ComponentModel.Composition.ExportAttribute";
+        };
+
         // NOTE: we recognize the attribute by name, not precise type. This makes the generator 
         // more flexible and avoids requiring any sort of run-time dependency.
         var services = types
-            .Where(x => x.GetAttributes().Any(selector))
-            .Select((x, _) => new
+            .Select((x, _) =>
             {
-                Type = x,
-                Lifetime = (int)x.GetAttributes().First(selector).ConstructorArguments[0].Value!
-            });
+                var attrs = x.GetAttributes();
+                var serviceAttr = attrs.FirstOrDefault(IsService);
+                var service = serviceAttr != null || attrs.Any(IsExport);
+
+                if (!service)
+                    return null;
+
+                // Default lifetime is singleton for [Service], Transient for MEF
+                var lifetime = serviceAttr != null ? 0 : 2;
+                if (serviceAttr != null)
+                {
+                    lifetime = (int)serviceAttr.ConstructorArguments[0].Value!;
+                }
+                else
+                {
+                    // In NuGet MEF, [Shared] makes exports singleton
+                    if (attrs.Any(a => a.AttributeClass?.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat) == "global::System.Composition.SharedAttribute"))
+                    {
+                        lifetime = 0;
+                    }
+                    // In .NET MEF, [PartCreationPolicy(CreationPolicy.Shared)] does it.
```

Comment:
unfortunately, in MEF it isn't that straightforward... whether the part is a singleton or not depends on both the Export and the Import. Nevertheless in practice, most of the times the CreationPolicy will just be Any, and if the injected object is also a singleton, then the import will also be one: https://learn.microsoft.com/en-us/dotnet/framework/mef/attributed-programming-model-overview-mef#creation-policies

Not sure what to recommend for the default behavior in this codegen, but if I were to look at our own usages, it's mostly singletons except for very few cases where we explicitly set NonShared....

Reply:
Sure, but I don't care much for MEF. This is just a "nice to have" thing, for convenience. In my (mostly VS) experience, things are almost always singletons, annotated with `[PartCreationPolicy(Shared)]`. There is no `[NotShared]` in System.Composition nor System.ComponentModel.Composition, which is why I made that the default behavior for the generator: you can only tell MEF that something *is* shared, so it made sense to make non-shared (transient) the default when no such attribute is present.